HW0/P5/P5.txt

```diff
@@ -1 +1,23 @@
-# Your answers here.
+Problem 5.
+
+Answers:
+1. 255 s
+2. 34 s
+3. 8 s
+4. ceiling(log2(N)) or N-1 (see attached)
+5. 15s
+6. a) 260 s b) 133 s c) Give 16 bags so as to balance out the gains from serial computations due to the lag in process distribution with the power of parallelization (35 s).
+
+Explanations:
+1. Addition is commutative and associative, and addition requires two "bags" to be processed at once. Therefore, the cashier must add 256 - 1 = 255 times to obtain the total.
+2. 8 workers together can process 16 bags per second. To process 256 bags, it will require 16 seconds; the 128 sums can be summed in 8 seconds, the resulting 64 sums can be processed in 4 seconds, and the following 32 sums in 2 seconds, yielding 16 sums requiring an additional 1 second to be further combined to 8 sums, in a binary tree fashion. However, further computations will be unable to take advantage of the available labor, as future computations require the present computations to complete first. Thus, each further summation will be an additional 1 second, halving the number of "sums" until the total sum remains (thus, 16 + 8 + 4 + 2 + 1 + 1 + 1 + 1 = 34 s)
+3. All 256 bags could be used by a maximum of 128 employees in the first "round" of computation, yielding 128 numbers to be added up. Then, in the second "round" of computation, 64 sums would be maximally obtained. In a binary tree fashion, we would see a halving of calculations needing to be done at each step. Thus, 8 rounds of computation would be required to analyze all 256 bags (2^8 = 256).
+4. With a countably "infinite" number of employees, the time required would be, in seconds, a ceiling of log base 2 of the number of bags, whereas a single employee would require N - 1 seconds.
+5. Going back to #3, neglecting communication delay, it requires 8 s to complete this task. Now, with communication delay, each round of computation requires a round of cross-talk between cashiers, which means that there is a 1 s delay between each of the 8 stages. Therefore, there is an additional 7 s delay on top of the 8 s = 15 s.
+6. It will take 256 s to distribute 1 bag to 1 cashier, so let's look at the terminating stages: once all 256 cashiers have received 1 bag, the last two need to (1) talk, (2) sum, (3) communicate with the rest of the "sum", and (4) sum for the final value. Hence, 256 + 4 s = 260 s.
```

Comment:
Yes, because the time it takes for two cashiers with one bag each (e.g., beginning after 2 seconds)  to sum (1 s to communicate, 1 s to add) is the same time it takes for you to give the two bags (2 s), one to each cashier. Thus, by the time the 256th bag is given and the 255th and 256th cashiers have communicated their bags and the 256th cashier has summed, the 254th cashier will already be poised with the cumulative sum from all other 254 cashiers (thus, only 2 s more = 2 + 2 + 256 = 260 s). Try drawing it out if you aren't convinced (start with 8 bags).